timemachine/fe/mcgregor.py

```diff
@@ -244,13 +167,183 @@ def to_networkx(self) -> nx.Graph:
+        )
+
+        return Marcs(marcs)
+
+    def refine(self, g1: Graph, g2: Graph, new_v1: int, new_v2: int) -> "Marcs":
+        """Set to False entries in marcs corresponding to edge pairs that can no longer correspond under the addition of
+        (new_v1, new_v2) to the mapping"""
+
+        new_marcs = np.array(self.marcs)
+        e1 = g1.get_edges_as_vector(new_v1)
+
+        if new_v2 == UNMAPPED:
+            # Set to False rows corresponding to the edges of new_v1
+            new_marcs[e1] = False
+        else:
+            e2 = g2.get_edges_as_vector(new_v2)
+
+            # Set to False edge pairs (e1, e2) where either
+            # 1. e1 connects to new_v1 but e2 does NOT connect to new_v2
+            # 2. e1 does NOT connect to new_v1 but e2 connects to new_v2
+            mask = e1[:, None] == e2[None, :]
+            new_marcs &= mask
+
+        return Marcs(new_marcs)
+
+
+@dataclass(frozen=True)
+class AtomMap:
+    a_to_b: Tuple[int, ...]
+    b_to_a: Tuple[int, ...]
+
+    @classmethod
+    def init(cls, n_1: int, n_2: int) -> "AtomMap":
+        return cls((UNMAPPED,) * n_1, (UNMAPPED,) * n_2)
+
+    def add(self, new_v1: int, new_v2: int) -> "AtomMap":
+        def set_at(xs: Tuple[int, ...], idx: int, val: int) -> Tuple[int, ...]:
+            return xs[:idx] + (val,) + xs[idx + 1 :]
+
+        return AtomMap(
+            set_at(self.a_to_b, new_v1, new_v2),
+            set_at(self.b_to_a, new_v2, new_v1),
+        )
+
+    @cached_property
+    def core_size(self):
+        return sum(1 for j in self.a_to_b if j != UNMAPPED)
+
+
+def _verify_map_preserves_core_edges(g1: Graph, g2: Graph, new_v1: int, new_v2: int, atom_map: AtomMap) -> bool:
+    def verify(g1: Graph, g2: Graph, new_v1: int, map_1_to_2: Sequence[int]):
+        for e1 in g1.get_edges(new_v1):
+            src, dst = g1.edges[e1]
+            src_2, dst_2 = map_1_to_2[src], map_1_to_2[dst]
+            if src_2 != UNMAPPED and dst_2 != UNMAPPED:
+                # both ends are mapped
+                # see if this edge is present in g2
+                if not g2.cmat[src_2][dst_2]:
+                    return False
+        return True
+
+    return verify(g1, g2, new_v1, atom_map.a_to_b) and verify(g2, g1, new_v2, atom_map.b_to_a)
+
+
+@dataclass(frozen=True)
+class Node:
+    atom_map: AtomMap
+    marcs: Marcs
+    layer: int
+
+    @classmethod
+    def init(cls, g1: Graph, g2: Graph, predicate: NDArray[np.bool_]) -> "Node":
+        return cls(AtomMap.init(g1.n_vertices, g2.n_vertices), Marcs.from_predicate(g1, g2, predicate), 0)
+
+    def add(self, g1: Graph, g2: Graph, new_v2: int) -> "Node":
+        return Node(
+            self.atom_map.add(self.layer, new_v2),
+            self.marcs.refine(g1, g2, self.layer, new_v2),
+            self.layer + 1,
+        )
+
+    def skip(self, g1: Graph, g2: Graph) -> "Node":
+        return Node(
+            self.atom_map,
+            self.marcs.refine(g1, g2, self.layer, UNMAPPED),
```

Comment:
IIRC there was one optimization I'd forgotten to incorporate earlier on but might be easy to do now:

We can tell when a node in `g1` is `UNMAPPED` because we've skipped past the current `layer`. However, we don't do any marking of nodes in `g2` to the `UNMAPPED` state - but this can be easily determined by checking the "missing" nodes from the union of the priority idxs, as in we can set:

`[self.marcs.refine(g1, g2, UNMAPPED, v2) for v2 in nodes_missing_from_priority_idxs]`

Reply:
I played around with this and did see a reduction in nodes visited for some examples, but the effect wasn't dramatic enough to offset the cost of my (naive, unoptimized) implementation. I want to explore this more, but will defer to a future PR.

Added TODO comments in https://github.com/proteneer/timemachine/pull/1415/commits/e409bf2a6179eb22930e5f1ca1bcbf82bc072a6f.